From 0741b3e4d8e16be22f82996680ba7ef2f4c7df51 Mon Sep 17 00:00:00 2001
Message-Id: <0741b3e4d8e16be22f82996680ba7ef2f4c7df51.1421709247.git.chang-joon.lee@intel.com>
In-Reply-To: <3650eb44734ad4fbe098313349da8185595a20d8.1421709247.git.chang-joon.lee@intel.com>
References: <3650eb44734ad4fbe098313349da8185595a20d8.1421709247.git.chang-joon.lee@intel.com>
From: Arun R Murthy <arun.r.murthy@intel.com>
Date: Mon, 13 Oct 2014 15:14:02 +0530
Subject: [PATCH 60/75] MUST_REBASE [VPG]: adf/intel/vlv: program the
 watermarks

For now we have programmed the watermarks with a constant value and
trickle feed is always disabled.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed

For: GMINL-3547
Change-Id: I06e3df486e86f7b310e14a0742300181ae67ba66
Signed-off-by: Arun R Murthy <arun.r.murthy@intel.com>
Signed-off-by: Shobhit Kumar <shobhit.kumar@intel.com>
---
 drivers/video/adf/intel/core/vlv/vlv_mode_set.c    |   34 ++++++++++++++++
 .../video/adf/intel/include/core/vlv/vlv_dc_regs.h |   42 +++++++++++++++++++-
 2 files changed, 75 insertions(+), 1 deletion(-)

diff --git a/drivers/video/adf/intel/core/vlv/vlv_mode_set.c b/drivers/video/adf/intel/core/vlv/vlv_mode_set.c
index b8871f9..f94459b 100644
--- a/drivers/video/adf/intel/core/vlv/vlv_mode_set.c
+++ b/drivers/video/adf/intel/core/vlv/vlv_mode_set.c
@@ -71,6 +71,35 @@ void vlv_wait_for_pipe_off(int pipe)
 		pr_err("ADF: %s: pipe_off wait timed out\n", __func__);
 }
 
+void vlv_program_wm(void)
+{
+	REG_WRITE(DSPFW1,
+		   (DSPFW_SR_VAL << DSPFW_SR_SHIFT) |
+		   (DSPFW_CURSORB_VAL << DSPFW_CURSORB_SHIFT) |
+		   (DSPFW_PLANEB_VAL << DSPFW_PLANEB_SHIFT) |
+		   DSPFW_PLANEA_VAL);
+	REG_WRITE(DSPFW2,
+		   (DSPFW2_RESERVED) |
+		   (DSPFW_CURSORA_VAL << DSPFW_CURSORA_SHIFT) |
+		   DSPFW_PLANEC_VAL);
+	REG_WRITE(DSPFW3,
+		   (REG_READ(DSPFW3) & ~DSPFW_CURSOR_SR_MASK) |
+		   (DSPFW3_VLV));
+	REG_WRITE(DSPFW4, (DSPFW4_SPRITEB_VAL << DSPFW4_SPRITEB_SHIFT) |
+			(DSPFW4_CURSORA_VAL << DSPFW4_CURSORA_SHIFT) |
+			DSPFW4_SPRITEA_VAL);
+	REG_WRITE(DSPFW5, (DSPFW5_DISPLAYB_VAL << DSPFW5_DISPLAYB_SHIFT) |
+			(DSPFW5_DISPLAYA_VAL << DSPFW5_DISPLAYA_SHIFT) |
+			(DSPFW5_CURSORB_VAL << DSPFW5_CURSORB_SHIFT) |
+			DSPFW5_CURSORSR_VAL);
+	REG_WRITE(DSPFW6, DSPFW6_DISPLAYSR_VAL);
+	REG_WRITE(DSPFW7, (DSPFW7_SPRITED1_VAL << DSPFW7_SPRITED1_SHIFT) |
+			(DSPFW7_SPRITED_VAL << DSPFW7_SPRITED_SHIFT) |
+			(DSPFW7_SPRITEC1_VAL << DSPFW7_SPRITEC1_SHIFT) |
+			DSPFW7_SPRITEC_VAL);
+	REG_WRITE(DSPARB, VLV_DEFAULT_DSPARB);
+}
+
 int vlv_display_on(struct intel_pipe *pipe)
 {
 	struct dsi_pipe *dsi_pipe = NULL;
@@ -135,6 +164,11 @@ int vlv_display_on(struct intel_pipe *pipe)
 	/* Enable DPST */
 	vlv_dpst_display_on();
 
+	/* Program the watermarks */
+	vlv_program_wm();
+	/* Trickle feed is disabled by default */
+	REG_WRITE(MI_ARB_VLV, 0x00);
+
 	return 0;
 }
 
diff --git a/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h b/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h
index d51a668..8bfe19b 100644
--- a/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h
+++ b/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h
@@ -3316,13 +3316,14 @@
 #define   PLANEA_INVALID_GTT_STATUS		(1<<0)
 #define   DPINVGTT_STATUS_MASK			0xff
 
-#define DSPARB			0x70030
+#define DSPARB			(VLV_DISPLAY_BASE + 0x70030)
 #define   DSPARB_CSTART_MASK	(0x7f << 7)
 #define   DSPARB_CSTART_SHIFT	7
 #define   DSPARB_BSTART_MASK	(0x7f)
 #define   DSPARB_BSTART_SHIFT	0
 #define   DSPARB_BEND_SHIFT	9 /* on 855 */
 #define   DSPARB_AEND_SHIFT	0
+#define	VLV_DEFAULT_DSPARB	0xc080c080
 
 #define DSPFW1			(VLV_DISPLAY_BASE + 0x70034)
 #define   DSPFW_SR_SHIFT	23
@@ -3332,19 +3333,58 @@
 #define   DSPFW_PLANEB_SHIFT	8
 #define   DSPFW_PLANEB_MASK	(0x7f<<8)
 #define   DSPFW_PLANEA_MASK	(0x7f)
+#define   DSPFW1_VLV		0x3F8F0F0F
+#define   DSPFW_PLANEA_VAL	0x0F
+#define   DSPFW_PLANEB_VAL	0x0F
+#define   DSPFW_CURSORB_VAL	0x0F
+#define   DSPFW_SR_VAL		0x7F
 #define DSPFW2			(VLV_DISPLAY_BASE + 0x70038)
 #define   DSPFW_CURSORA_MASK	0x00003f00
 #define   DSPFW_CURSORA_SHIFT	8
 #define   DSPFW_PLANEC_MASK	(0x7f)
+#define   DSPFW2_VLV		0x0B0F0F0F
+#define   DSPFW_PLANEC_VAL	0x0F
+#define   DSPFW_CURSORA_VAL	0x0F
+#define   DSPFW2_RESERVED	(0xB0F<<16)
 #define DSPFW3			(VLV_DISPLAY_BASE + 0x7003c)
 #define   DSPFW_HPLL_SR_EN	(1<<31)
 #define   DSPFW_CURSOR_SR_SHIFT	24
+#define   DSPFW3_VLV			0x0
 #define   PINEVIEW_SELF_REFRESH_EN	(1<<30)
 #define   DSPFW_CURSOR_SR_MASK		(0x3f<<24)
 #define   DSPFW_HPLL_CURSOR_SHIFT	16
 #define   DSPFW_HPLL_CURSOR_MASK	(0x3f<<16)
 #define   DSPFW_HPLL_SR_MASK		(0x1ff)
 #define DSPFW4			(VLV_DISPLAY_BASE + 0x70070)
+#define   DSPFW4_SPRITEA_VAL	0x04
+#define   DSPFW4_CURSORA_SHIFT	8
+#define   DSPFW4_CURSORA_VAL	0x04
+#define   DSPFW4_SPRITEB_SHIFT	16
+#define   DSPFW4_SPRITEB_VAL	0x04
+#define   DSPFW4_VLV		0x00040404
+#define DSPFW5			(VLV_DISPLAY_BASE + 0x70074)
+#define   DSPFW5_CURSORSR_VAL	0x4
+#define   DSPFW5_CURSORB_SHIFT	8
+#define   DSPFW5_CURSORB_VAL	0x4
+#define   DSPFW5_DISPLAYA_SHIFT	16
+#define   DSPFW5_DISPLAYA_VAL	0x4
+#define   DSPFW5_DISPLAYB_SHIFT	24
+#define   DSPFW5_DISPLAYB_VAL	0x4
+#define   DSPFW5_VLV		0x04040404
+#define DSPFW6			(VLV_DISPLAY_BASE + 0x70078)
+#define   DSPFW6_DISPLAYSR_VAL	0xF
+#define   DSPFW6_VLV		0x0000000F
+#define DSPFW7			(VLV_DISPLAY_BASE + 0x7007c)
+#define   DSPFW7_SPRITEC_VAL	0x0F
+#define   DSPFW7_SPRITEC1_VAL	0x04
+#define   DSPFW7_SPRITEC1_SHIFT	8
+#define   DSPFW7_SPRITED_VAL	0x0F
+#define   DSPFW7_SPRITED_SHIFT	16
+#define   DSPFW7_SPRITED1_VAL	0x04
+#define   DSPFW7_SPRITED1_SHIFT	24
+#define   DSPFW7_VLV		0x040F040F
+
+#define DSPFW4			(VLV_DISPLAY_BASE + 0x70070)
 #define DSPFW7			(VLV_DISPLAY_BASE + 0x7007c)
 
 /* drain latency register values*/
-- 
1.7.9.5

